typing_extensions/src_py3/typing_extensions.py

```diff
@@ -2166,3 +2169,226 @@ class TypeAlias(metaclass=_TypeAliasMeta, _root=True):
         It's invalid when used anywhere except as in the example above.
         """
         __slots__ = ()
+
+
+# Python 3.10+ has PEP 612
+if hasattr(typing, 'ParamSpec'):
+    ParamSpec = typing.ParamSpec
+else:
+    # Inherits from list as a workaround for Callable checks in Python < 3.9.2.
+    class ParamSpec(list):
+        """Parameter specification variable.
+
+        Usage::
+
+           P = ParamSpec('P')
+
+        Parameter specification variables exist primarily for the benefit of static
+        type checkers.  They are used to forward the parameter types of one
+        callable to another callable, a pattern commonly found in higher order
+        functions and decorators.  They are only valid when used in ``Concatenate``,
+        or s the first argument to ``Callable``. In Python 3.10 and higher,
+        they are also supported in user-defined Generics at runtime.
+        See class Generic for more information on generic types.  An
+        example for annotating a decorator::
+
+           T = TypeVar('T')
+           P = ParamSpec('P')
+
+           def add_logging(f: Callable[P, T]) -> Callable[P, T]:
+               '''A type-safe decorator to add logging to a function.'''
+               def inner(*args: P.args, **kwargs: P.kwargs) -> T:
+                   logging.info(f'{f.__name__} was called')
+                   return f(*args, **kwargs)
+               return inner
+
+           @add_logging
+           def add_two(x: float, y: float) -> float:
+               '''Add two numbers together.'''
+               return x + y
+
+        Parameter specification variables defined with covariant=True or
+        contravariant=True can be used to declare covariant or contravariant
+        generic types.  These keyword arguments are valid, but their actual semantics
+        are yet to be decided.  See PEP 612 for details.
+
+        Parameter specification variables can be introspected. e.g.:
+
+           P.__name__ == 'T'
+           P.__bound__ == None
+           P.__covariant__ == False
+           P.__contravariant__ == False
+
+        Note that only parameter specification variables defined in global scope can
+        be pickled.
+        """
+        args = object()
```

Comment:
I'm a bit unhappy with this because it makes the runtime annotations just look like this:
```
In [11]: takes_int_str.__annotations__
Out[11]: 
{'args': <object at 0x7ff1831d5f30>,
 'kwargs': <object at 0x7ff1831d5f60>,
 'return': ~R}
```
However, it's the exact same in CPython, so it makes sense to follow the same principle here. I'll plan to submit a change to CPython to make this instead return some helper object pointing back to the ParamSpec; if that gets accepted we can apply the same change here.

For now no change is required here.

Reply:
This was definitely something I overlooked when I implemented the CPython version :(. How I'd do it in typing.py:

1. Create a `_ParamSpecGenericAlias`.
2.  `__origin__` is the ParamSpec it belongs to, `__args__` is just a tuple with 1 element - "args" or "kwargs".
3. Override the `__repr__` to print `~T.(args|kwargs)` instead of the default `~T[args]` found in genericaliases.

I look forward to seeing your PR!

Reply:
Yes, I was thinking something similar, except with different classes for args and kwargs, but either way should work.